```
error[E0004]: non-exhaustive patterns: `space` and `round` not covered

    --> /home/travis/build/servo/servo/target/debug/build/style-b398ea74b908472c/out/gecko_properties.rs:3222:42

     |

3222 |         let (repeat_x, repeat_y) = match servo {

     |                                          ^^^^^ patterns `space` and `round` not covered

error: aborting due to previous error

error: Could not compile `style`.

To learn more, run the command again with --verbose.

The command "./mach test-geckolib" exited with 101.
```

--- a/components/gfx/display_list/mod.rs
+++ b/components/gfx/display_list/mod.rs
@@ -1148,6 +1148,10 @@ pub struct ImageDisplayItem {
     /// direction to tile the entire bounds.
     pub stretch_size: Size2D<Au>,
 
+    /// The amount of space to add to the right and bottom part of each tile, when the image
+    /// is tiled.
+    pub tile_spacing: Size2D<Au>,
+
     /// The algorithm we should use to stretch the image. See `image_rendering` in CSS-IMAGES-3 §
     /// 5.3.
     pub image_rendering: image_rendering::T,
@@ -1360,6 +1364,7 @@ impl DisplayItem {
                 paint_context.draw_image(
                     &image_item.base.bounds,
                     &image_item.stretch_size,
+                    &image_item.tile_spacing,
                     &image_item.webrender_image,
                     &image_item.image_data
                                .as_ref()

--- a/components/gfx/paint_context.rs
+++ b/components/gfx/paint_context.rs
@@ -189,6 +189,7 @@ impl<'a> PaintContext<'a> {
     pub fn draw_image(&self,
                       bounds: &Rect<Au>,
                       stretch_size: &Size2D<Au>,
+                      tile_spacing: &Size2D<Au>,
                       image_info: &WebRenderImageInfo,
                       image_data: &[u8],
                       image_rendering: image_rendering::T) {
@@ -229,7 +230,7 @@ impl<'a> PaintContext<'a> {
         let draw_options = DrawOptions::new(1.0, CompositionOp::Over, AntialiasMode::None);
 
         // Fast path: No need to create a pattern.
-        if bounds.size == *stretch_size {
+        if bounds.size == *stretch_size && *tile_spacing == Size2D::zero() {
             draw_target_ref.draw_surface(azure_surface,
                                          dest_rect,
                                          source_rect,
@@ -243,8 +244,8 @@ impl<'a> PaintContext<'a> {
         // Annoyingly, surface patterns in Azure/Skia are relative to the top left of the *canvas*,
         // not the rectangle we're drawing to. So we need to translate it explicitly.
         let matrix = Matrix2D::identity().pre_translated(dest_rect.origin.x, dest_rect.origin.y);
-        let stretch_size = stretch_size.to_nearest_azure_size(scale);
-        if source_rect.size == stretch_size {
+        let azure_stretch_size = stretch_size.to_nearest_azure_size(scale);
+        if source_rect.size == azure_stretch_size && *tile_spacing == Size2D::zero() {
             let pattern = SurfacePattern::new(azure_surface.azure_source_surface,
                                               true,
                                               true,
@@ -258,10 +259,11 @@ impl<'a> PaintContext<'a> {
         // Slow path: Both stretch and a pattern are needed.
         let draw_surface_options = DrawSurfaceOptions::new(draw_surface_filter, true);
         let draw_options = DrawOptions::new(1.0, CompositionOp::Over, AntialiasMode::None);
+        let temporary_target_size = (*stretch_size + *tile_spacing).to_nearest_azure_size(scale);
         let temporary_draw_target =
-            self.draw_target.create_similar_draw_target(&stretch_size.to_azure_int_size(),
+            self.draw_target.create_similar_draw_target(&temporary_target_size.to_azure_int_size(),
                                                         self.draw_target.get_format());
-        let temporary_dest_rect = Rect::new(Point2D::new(0.0, 0.0), stretch_size);
+        let temporary_dest_rect = Rect::new(Point2D::new(0.0, 0.0), azure_stretch_size);
         temporary_draw_target.draw_surface(azure_surface,
                                            temporary_dest_rect,
                                            source_rect,

--- a/components/layout/display_list_builder.rs
+++ b/components/layout/display_list_builder.rs
@@ -547,46 +547,74 @@ impl FragmentDisplayListBuilding for Fragment {
             let position = *get_cyclic(&background.background_position.0, index);
             // Use `background-position` to get the offset.
             let horizontal_position = model::specified(position.horizontal,
-                                                     bounds.size.width - image_size.width);
+                                                       bounds.size.width - image_size.width);
             let vertical_position = model::specified(position.vertical,
-                                                   bounds.size.height - image_size.height);
+                                                     bounds.size.height - image_size.height);
 
-            let abs_x = border.left + virtual_origin_x + horizontal_position + origin_x;
-            let abs_y = border.top + virtual_origin_y + vertical_position + origin_y;
+            // The anchor position for this background, based on both the background-attachment
+            // and background-position properties.
+            let anchor_origin_x = border.left + virtual_origin_x + origin_x + horizontal_position;
+            let anchor_origin_y = border.top + virtual_origin_y + origin_y + vertical_position;
+
+            let mut tile_spacing = Size2D::zero();
+            let mut stretch_size = image_size;
 
             // Adjust origin and size based on background-repeat
             match *get_cyclic(&background.background_repeat.0, index) {
                 background_repeat::single_value::T::no_repeat => {
-                    bounds.origin.x = abs_x;
-                    bounds.origin.y = abs_y;
+                    bounds.origin.x = anchor_origin_x;
+                    bounds.origin.y = anchor_origin_y;
                     bounds.size.width = image_size.width;
                     bounds.size.height = image_size.height;
                 }
                 background_repeat::single_value::T::repeat_x => {
-                    bounds.origin.y = abs_y;
+                    bounds.origin.y = anchor_origin_y;
                     bounds.size.height = image_size.height;
                     ImageFragmentInfo::tile_image(&mut bounds.origin.x,
                                                   &mut bounds.size.width,
-                                                  abs_x,
-                                                  image_size.width.to_nearest_px() as u32);
+                                                  anchor_origin_x,
+                                                  image_size.width);
                 }
                 background_repeat::single_value::T::repeat_y => {
-                    bounds.origin.x = abs_x;
+                    bounds.origin.x = anchor_origin_x;
                     bounds.size.width = image_size.width;
                     ImageFragmentInfo::tile_image(&mut bounds.origin.y,
                                                   &mut bounds.size.height,
-                                                  abs_y,
-                                                  image_size.height.to_nearest_px() as u32);
+                                                  anchor_origin_y,
+                                                  image_size.height);
                 }
                 background_repeat::single_value::T::repeat => {
                     ImageFragmentInfo::tile_image(&mut bounds.origin.x,
-                                                &mut bounds.size.width,
-                                                abs_x,
-                                                image_size.width.to_nearest_px() as u32);
+                                                  &mut bounds.size.width,
+                                                  anchor_origin_x,
+                                                  image_size.width);
                     ImageFragmentInfo::tile_image(&mut bounds.origin.y,
-                                                &mut bounds.size.height,
-                                                abs_y,
-                                                image_size.height.to_nearest_px() as u32);
+                                                  &mut bounds.size.height,
+                                                  anchor_origin_y,
+                                                  image_size.height);
+                }
+                background_repeat::single_value::T::space => {
+                    ImageFragmentInfo::tile_image_spaced(&mut bounds.origin.x,
+                                                         &mut bounds.size.width,
+                                                         &mut tile_spacing.width,
+                                                         anchor_origin_x,
+                                                         image_size.width);
+                    ImageFragmentInfo::tile_image_spaced(&mut bounds.origin.y,
+                                                         &mut bounds.size.height,
+                                                         &mut tile_spacing.height,
+                                                         anchor_origin_y,
+                                                         image_size.height);
+
+                }
+                background_repeat::single_value::T::round => {
+                    ImageFragmentInfo::tile_image_round(&mut bounds.origin.x,
+                                                        &mut bounds.size.width,
+                                                        anchor_origin_x,
+                                                        &mut stretch_size.width);
+                    ImageFragmentInfo::tile_image_round(&mut bounds.origin.y,
+                                                        &mut bounds.size.height,
+                                                        anchor_origin_y,
+                                                        &mut stretch_size.height);
                 }
             };
 
@@ -600,7 +628,8 @@ impl FragmentDisplayListBuilding for Fragment {
               base: base,
               webrender_image: webrender_image,
               image_data: image_data.map(Arc::new),
-              stretch_size: Size2D::new(image_size.width, image_size.height),
+              stretch_size: stretch_size,
+              tile_spacing: tile_spacing,
               image_rendering: style.get_inheritedbox().image_rendering.clone(),
             }));
 
@@ -1257,6 +1286,7 @@ impl FragmentDisplayListBuilding for Fragment {
                         webrender_image: WebRenderImageInfo::from_image(image),
                         image_data: Some(Arc::new(image.bytes.clone())),
                         stretch_size: stacking_relative_content_box.size,
+                        tile_spacing: Size2D::zero(),
                         image_rendering: self.style.get_inheritedbox().image_rendering.clone(),
                     }));
                 }
@@ -1300,6 +1330,7 @@ impl FragmentDisplayListBuilding for Fragment {
                                     key: canvas_data.image_key,
                                 },
                                 stretch_size: stacking_relative_content_box.size,
+                                tile_spacing: Size2D::zero(),
                                 image_rendering: image_rendering::T::Auto,
                             })
                         }

--- a/components/layout/fragment.rs
+++ b/components/layout/fragment.rs
@@ -425,18 +425,73 @@ impl ImageFragmentInfo {
         }
     }
 
+    pub fn tile_image_round(position: &mut Au,
+                            size: &mut Au,
+                            absolute_anchor_origin: Au,
+                            image_size: &mut Au) {
+        if *size == Au(0) || *image_size == Au(0) {
+            *position = Au(0);
+            *size =Au(0);
+            return;
+        }
+
+        let number_of_tiles = (size.to_f32_px() / image_size.to_f32_px()).round().max(1.0);
+        *image_size = *size / (number_of_tiles as i32);
+        ImageFragmentInfo::tile_image(position, size, absolute_anchor_origin, *image_size);
+    }
+
+    pub fn tile_image_spaced(position: &mut Au,
+                             size: &mut Au,
+                             tile_spacing: &mut Au,
+                             absolute_anchor_origin: Au,
+                             image_size: Au) {
+        if *size == Au(0) || image_size == Au(0) {
+            *position = Au(0);
+            *size = Au(0);
+            *tile_spacing = Au(0);
+            return;
+        }
+
+        // Per the spec, if the space available is not enough for two images, just tile as
+        // normal but only display a single tile.
+        if image_size * 2 >= *size {
+            ImageFragmentInfo::tile_image(position,
+                                          size,
+                                          absolute_anchor_origin,
+                                          image_size);
+            *tile_spacing = Au(0);
+            *size = image_size;;
+            return;
+        }
+
+        // Take the box size, remove room for two tiles on the edges, and then calculate how many
+        // other tiles fit in between them.
+        let size_remaining = *size - (image_size * 2);
+        let num_middle_tiles = (size_remaining.to_f32_px() / image_size.to_f32_px()).floor() as i32;
+
+        // Allocate the remaining space as padding between tiles. background-position is ignored
+        // as per the spec, so the position is just the box origin. We are also ignoring
+        // background-attachment here, which seems unspecced when combined with
+        // background-repeat: space.
+        let space_for_middle_tiles = image_size * num_middle_tiles;
+        *tile_spacing = (size_remaining - space_for_middle_tiles) / (num_middle_tiles + 1);
+    }
+
     /// Tile an image
-    pub fn tile_image(position: &mut Au, size: &mut Au, virtual_position: Au, image_size: u32) {
+    pub fn tile_image(position: &mut Au,
+                      size: &mut Au,
+                      absolute_anchor_origin: Au,
+                      image_size: Au) {
         // Avoid division by zero below!
-        let image_size = image_size as i32;
-        if image_size == 0 {
+        if image_size == Au(0) {
             return
         }
 
-        let delta_pixels = (virtual_position - *position).to_px();
-        let tile_count = (delta_pixels + image_size - 1) / image_size;
-        let offset = Au::from_px(image_size * tile_count);
-        let new_position = virtual_position - offset;
+        let delta_pixels = absolute_anchor_origin - *position;
+        let image_size_px = image_size.to_f32_px();
+        let tile_count = ((delta_pixels.to_f32_px() + image_size_px - 1.0) / image_size_px).floor();
+        let offset = image_size * (tile_count as i32);
+        let new_position = absolute_anchor_origin - offset;
         *size = *position - new_position + *size;
         *position = new_position;
     }

--- a/components/layout/webrender_helpers.rs
+++ b/components/layout/webrender_helpers.rs
@@ -427,12 +427,10 @@ impl WebRenderDisplayItemConverter for DisplayItem {
                 if let Some(id) = item.webrender_image.key {
                     if item.stretch_size.width > Au(0) &&
                        item.stretch_size.height > Au(0) {
-                        // TODO(gw): Pass through the tile spacing once the other
-                        //           changes related to this land (parsing etc).
                         builder.push_image(item.base.bounds.to_rectf(),
                                            item.base.clip.to_clip_region(frame_builder),
                                            item.stretch_size.to_sizef(),
-                                           Size2D::zero(),
+                                           item.tile_spacing.to_sizef(),
                                            item.image_rendering.to_image_rendering(),
                                            id);
                     }

--- a/components/style/properties/gecko.mako.rs
+++ b/components/style/properties/gecko.mako.rs
@@ -1011,14 +1011,20 @@ fn static_assert() {
         use gecko_bindings::structs::nsStyleImageLayers_Repeat;
         use gecko_bindings::structs::NS_STYLE_IMAGELAYER_REPEAT_REPEAT;
         use gecko_bindings::structs::NS_STYLE_IMAGELAYER_REPEAT_NO_REPEAT;
+        use gecko_bindings::structs::NS_STYLE_IMAGELAYER_REPEAT_SPACE;
+        use gecko_bindings::structs::NS_STYLE_IMAGELAYER_REPEAT_ROUND;
 
         let (repeat_x, repeat_y) = match servo {
           T::repeat_x => (NS_STYLE_IMAGELAYER_REPEAT_REPEAT,
                           NS_STYLE_IMAGELAYER_REPEAT_NO_REPEAT),
           T::repeat_y => (NS_STYLE_IMAGELAYER_REPEAT_NO_REPEAT,
                           NS_STYLE_IMAGELAYER_REPEAT_REPEAT),
-          T::repeat => (NS_STYLE_IMAGELAYER_REPEAT_REPEAT,
-                        NS_STYLE_IMAGELAYER_REPEAT_REPEAT),
+          T::repeat  => (NS_STYLE_IMAGELAYER_REPEAT_REPEAT,
+                         NS_STYLE_IMAGELAYER_REPEAT_REPEAT),
+          T::space => (NS_STYLE_IMAGELAYER_REPEAT_SPACE,
+                       NS_STYLE_IMAGELAYER_REPEAT_SPACE),
+          T::round => (NS_STYLE_IMAGELAYER_REPEAT_ROUND,
+                       NS_STYLE_IMAGELAYER_REPEAT_ROUND),
           T::no_repeat => (NS_STYLE_IMAGELAYER_REPEAT_NO_REPEAT,
                            NS_STYLE_IMAGELAYER_REPEAT_NO_REPEAT),
         };

--- a/components/style/properties/longhand/background.mako.rs
+++ b/components/style/properties/longhand/background.mako.rs
@@ -119,7 +119,7 @@ ${helpers.predefined_type("background-color", "CSSColor",
 </%helpers:vector_longhand>
 
 ${helpers.single_keyword("background-repeat",
-                         "repeat repeat-x repeat-y no-repeat",
+                         "repeat repeat-x repeat-y space round no-repeat",
                          vector=True,
                          animatable=False)}
 

--- a/components/style/properties/longhand/svg.mako.rs
+++ b/components/style/properties/longhand/svg.mako.rs
@@ -87,7 +87,7 @@ ${helpers.single_keyword("mask-mode",
 // TODO implement all of repeat-style for background and mask
 // https://drafts.csswg.org/css-backgrounds-3/#repeat-style
 ${helpers.single_keyword("mask-repeat",
-                         "repeat repeat-x repeat-y no-repeat",
+                         "repeat repeat-x repeat-y space round no-repeat",
                          vector=True,
                          products="gecko",
                          animatable=False)}

--- a/tests/wpt/metadata-css/css-backgrounds-3_dev/html4/background-repeat-round.htm.ini
+++ b/tests/wpt/metadata-css/css-backgrounds-3_dev/html4/background-repeat-round.htm.ini
@@ -1,3 +1,0 @@
-[background-repeat-round.htm]
-  type: reftest
-  expected: FAIL

--- a/tests/wpt/metadata-css/css-backgrounds-3_dev/html4/background-repeat-space.htm.ini
+++ b/tests/wpt/metadata-css/css-backgrounds-3_dev/html4/background-repeat-space.htm.ini
@@ -1,3 +1,0 @@
-[background-repeat-space.htm]
-  type: reftest
-  expected: FAIL

--- a/tests/wpt/metadata-css/css-backgrounds-3_dev/html4/background-size-025.htm.ini
+++ b/tests/wpt/metadata-css/css-backgrounds-3_dev/html4/background-size-025.htm.ini
@@ -1,3 +1,0 @@
-[background-size-025.htm]
-  type: reftest
-  expected: FAIL

